Only clear builder payment if the slashed validator is the proposer (#5365)
ethereum/consensus-specs: specs/gloas/beacon-chain.md

        assert bls.Verify(proposer.pubkey, signing_root, signed_header.signature)

    # [New in Gloas:EIP7732]
    # Remove the BuilderPendingPayment corresponding to this proposal if it is
    # still in the 2-epoch window. Only clear it when the slashed validator is

>     # the proposer that armed the payment; otherwise an unrelated same-slot

Reviewer: I don't like the word "armed" here. I would suggest this instead.
```suggestion
    # the proposer associated with the payment; otherwise an unrelated same-slot
```

Author: Reworded to "associated with" in b46bfe0. Thanks!

    # equivocation could grief an honest proposer's payment.
    slot = header_1.slot
    proposal_epoch = compute_epoch_at_slot(slot)
    if proposal_epoch == get_current_epoch(state):
        payment_index = SLOTS_PER_EPOCH + slot % SLOTS_PER_EPOCH
        payment = state.builder_pending_payments[payment_index]

>         if payment.proposer_index == header_1.proposer_index:

Reviewer 2: one thing that may make it slightly more readable and avoid a repetition is to first slash the validator, then add a single if proposer_index == header_1.proposer_index and return early if not. Then keep the previous behavior, minimizing the diff.

Author: Done! moved slash_validator to the top so the payment handling can return early: return if the slot is outside the 2-epoch window, return if the slashed validator isn't the payment's proposer, then the original clear. Removes the duplicated guard. It's behavior-preserving (slashing and the payment-clear touch disjoint state), so no test changes were needed -- full gloas fork suite + lint stayed green locally.

Reviewer: Yeah sorry I thought this would look better. I would prefer we revert https://github.com/ethereum/consensus-specs/pull/5365/commits/b46bfe073950a4b3722a375f85e30210c8ecb8d2.

Reviewer: Except for the "associated with" change 😅

Author: Done!

Reviewer 3: ```python
   if proposal_epoch == get_current_epoch(state):
        payment_index = SLOTS_PER_EPOCH + slot % SLOTS_PER_EPOCH
    elif proposal_epoch == get_previous_epoch(state):
        payment_index = slot % SLOTS_PER_EPOCH
    else:
        payment_index = None
    
    if payment_index is not None:
        payment = state.builder_pending_payments[payment_index]
        if payment.proposer_index == header_1.proposer_index:
            state.builder_pending_payments[payment_index] = BuilderPendingPayment()
```
FYI, this is another way.

Reviewer: IMO this is still more complicated than the current version.

Reviewer 3: I don't agree that this is more complicated (without the modifier comment you have in your mind) but I don't care whichever we do.

Reviewer 3: Just to be clear, I don't think either code is more complicated than the other meaningfully.

Author: Thanks both! Since it's approved and you both agree the two forms aren't meaningfully different, I'll keep the current version.
